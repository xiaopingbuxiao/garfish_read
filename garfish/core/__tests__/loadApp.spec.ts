import {
  __MockBody__,
  __MockHead__,
  __MockHtml__,
  appContainerId,
} from '@garfish/utils';
import { mockStaticServer } from '@garfish/test-suite';
import Garfish from '../src/index';
import assert from 'assert';

describe('Core: load process', () => {
  let GarfishInstance: Garfish;
  let container;

  const vueSubAppEntry = './resources/vueApp.html';
  const reactSubAppEntry = './resources/reactApp.html';
  const vue3AppRenderNode = 'hello-world-vue3';
  const vue3SubAppEntry = './resources/vue3App.html';
  const asyncProviderApp = './resources/asyncProviderApp.html';

  mockStaticServer({
    baseDir: __dirname,
  });

  beforeEach(() => {
    container = document.createElement('div');
    container.setAttribute('id', 'container');
    document.body.appendChild(container);
    GarfishInstance = new Garfish({});
  });

  afterEach(() => {
    document.body.removeChild(container);
  });

  it('domGetter cssSelector', async () => {
    const app = await GarfishInstance.loadApp('vue-app', {
      entry: vueSubAppEntry,
      domGetter: '#container',
    });

    assert(app, 'app should be loaded');
    await app.mount();

    const appContainer = container.querySelector(`[id^=${appContainerId}]`);
    expect(appContainer.childNodes[0]).toMatchSnapshot();
  });

  it('domGetter cssSelector delay 2000ms can work normally', async () => {
    const container = document.createElement('div');
    setTimeout(() => {
      container.setAttribute('id', 'async-container');
      document.body.appendChild(container);
    }, 1000);

    const app = await GarfishInstance.loadApp('vue-app', {
      entry: vueSubAppEntry,
      domGetter: '#async-container',
    });

    assert(app, 'app should be loaded');
    await app.mount();

    const appContainer = container.querySelector(`[id^=${appContainerId}]`);
    assert(appContainer, 'app container should be loaded');
    expect(appContainer.childNodes[0]).toMatchSnapshot();

    app.unmount();
    document.body.removeChild(container);
  });

  it('domGetter cssSelector delay 3500ms throw error', async () => {
    const container = document.createElement('div');
    setTimeout(() => {
      container.setAttribute('id', 'delay-async-container');
      document.body.appendChild(container);
    }, 3500);

    const app = await GarfishInstance.loadApp('vue-app', {
      entry: vueSubAppEntry,
      domGetter: '#delay-async-container',
    });

    assert(app, 'app should be loaded');
    await expect(app.mount.bind(app)).rejects.toThrowError(/Invalid domGetter/);
  });

  it('domGetter is function', async () => {
    const app = await GarfishInstance.loadApp('vue-app', {
      entry: vueSubAppEntry,
      domGetter: () => document.querySelector('#container'),
    });

    assert(app, 'app should be loaded');
    await app.mount();

    const appContainer = container.querySelector(`[id^=${appContainerId}]`);
    expect(appContainer.childNodes[0]).toMatchSnapshot();
  });

  it('load error app', async () => {
    const container = document.createElement('div');
    container.setAttribute('id', 'container');
    document.body.appendChild(container);
    const app = await GarfishInstance.loadApp('vue-app', {
      entry: reactSubAppEntry,
      domGetter: () => document.querySelector('#container'),
    });

    assert(app, 'app should be loaded');
    await expect(app.mount()).rejects.toThrowError('test msg');

    document.body.removeChild(container);
  });

  it("throw error when provide an 'name' opts in options", async () => {
    await expect(GarfishInstance.loadApp('vue-app')).rejects.toThrow();
  });

  it('get the options from register info by default', async () => {
    GarfishInstance.run({
      domGetter: '#container',
      apps: [
        {
          name: 'vue-app',
          entry: vueSubAppEntry,
        },
      ],
    });

    await expect(GarfishInstance.loadApp('vue-app')).resolves.toMatchObject({
      appInfo: {
        entry: vueSubAppEntry,
      },
    });
  });

  it('config will be deepMerged with global config、register app info and provided by options in loadApp', async () => {
    GarfishInstance.run({
      basename: 'demo',
      domGetter: '#container',
      apps: [
        {
          name: 'vue-app',
          entry: vueSubAppEntry,
        },
      ],
    });

    await expect(
      GarfishInstance.loadApp('vue-app', {
        domGetter: '#container',
        entry: vueSubAppEntry,
      }),
    ).resolves.toMatchObject({
      appInfo: {
        basename: 'demo',
        entry: vueSubAppEntry,
      },
    });
  });

  it('config provided by options in loadApp will have the highest priority after config merged', async () => {
    const mockBeforeLoad = jest.fn(() => {
      GarfishInstance.appInfos['vue-app'] = {
        name: 'vue-app',
        entry: vue3SubAppEntry,
      };
    });

    GarfishInstance.run({
      beforeLoad: mockBeforeLoad,
      apps: [
        {
          name: 'vue-app',
          entry: vueSubAppEntry,
          basename: '/vue',
        },
      ],
    });

    await expect(
      GarfishInstance.loadApp('vue-app', {
        domGetter: '#container',
      }),
    ).resolves.toMatchObject({
      appInfo: {
        entry: vue3SubAppEntry,
      },
    });

    const app = await GarfishInstance.loadApp('vue-app', {
      entry: vue3SubAppEntry,
      domGetter: '#container',
    });

    assert(app, 'app should be loaded');
    await app.mount();

    const appContainer = container.querySelectorAll(`[id^=${appContainerId}]`);
    expect(appContainer).toHaveLength(1);
    expect(appContainer[0].querySelectorAll(`[${__MockHtml__}]`)).toHaveLength(
      1,
    );
    expect(appContainer[0].querySelectorAll(`[${__MockHead__}]`)).toHaveLength(
      1,
    );
    expect(appContainer[0].querySelectorAll(`[${__MockBody__}]`)).toHaveLength(
      1,
    );
    expect(
      appContainer[0].querySelectorAll(`[id=${vue3AppRenderNode}]`),
    ).toHaveLength(1);
    app.unmount();
    expect(container.querySelectorAll(`[id^=${appContainerId}]`)).toHaveLength(
      0,
    );
  });

  it('provide an entry after beforeLoad if entry is not provided in register', async () => {
    const mockBeforeLoad = jest.fn(() => {
      GarfishInstance.appInfos['vue-app'] = {
        name: 'vue-app',
        entry: vueSubAppEntry,
      };
    });

    GarfishInstance.run({
      beforeLoad: mockBeforeLoad,
    });

    await expect(
      GarfishInstance.loadApp('vue-app', {
        domGetter: '#container',
      }),
    ).resolves.toMatchObject({
      appInfo: {
        entry: vueSubAppEntry,
      },
    });
  });

  it('throw error when entry is not provided after beforeLoad', async () => {
    await expect(
      GarfishInstance.loadApp('vue-app', {
        domGetter: '#container',
      }),
    ).rejects.toThrow(
      'Please provide the entry parameters or registered in advance of the app.',
    );
  });

  it('load the same app while return the cache instance if cache is true', async () => {
    await GarfishInstance.run({
      domGetter: '#container',
      apps: [
        {
          name: 'vue-app',
          entry: vueSubAppEntry,
          cache: true,
        },
      ],
    });

    const app1 = await GarfishInstance.loadApp('vue-app');
    assert(app1, 'app1 should be loaded');
    await app1.mount();
    const app2 = await GarfishInstance.loadApp('vue-app');
    assert(app2, 'app2 should be loaded');

    expect(app1.mounted).toBe(true);
    expect(app2.mounted).toEqual(app1.mounted);
  });

  it('load the same app while return the cache instance if cache is set true', async () => {
    await GarfishInstance.run({
      domGetter: '#container',
      apps: [
        {
          name: 'vue-app',
          entry: vueSubAppEntry,
          cache: false,
        },
      ],
    });

    const app1 = await GarfishInstance.loadApp('vue-app');
    assert(app1, 'app2 should be loaded');
    await app1.mount();
    const app2 = await GarfishInstance.loadApp('vue-app');
    assert(app2, 'app2 should be loaded');

    expect(app1.mounted).toBe(true);
    expect(app2.mounted).not.toEqual(app1.mounted);
  });

  it('load the same app while return the cache instance if cache is set false', async () => {
    GarfishInstance.run({
      domGetter: '#container',
      apps: [
        {
          name: 'vue-app',
          entry: vueSubAppEntry,
          cache: true,
        },
      ],
    });

    const app1 = await GarfishInstance.loadApp('vue-app');
    assert(app1, 'app2 should be loaded');
    await app1.mount();

    const app2 = await GarfishInstance.loadApp('vue-app', {
      entry: vue3SubAppEntry,
      cache: false,
    });
    assert(app2, 'app2 should be loaded');

    expect(app1.mounted).toBe(true);
    expect(app2.mounted).toBe(false);
    expect(app2.appInfo.entry).toEqual(vue3SubAppEntry);
  });

  it('destroy the app during rendering and then render again', async () => {
    GarfishInstance.run({
      domGetter: '#container',
      apps: [
        {
          name: 'async-provider',
          entry: asyncProviderApp,
          cache: true,
        },
      ],
    });

    const app = await GarfishInstance.loadApp('async-provider');
    assert(app, 'app2 should be loaded');
    expect(app.mounted).toBe(false);

    app.mount();
    app.unmount();

    await app.mount();
    expect(app.mounted).toBe(true);

    expect(document.body.contains(container)).toBe(true);
  });

  it('loadApp before registered, the `beforeLoad` hook will not receive the undefined app.', async () => {
    const mockBeforeLoad = jest.fn();

    GarfishInstance.run({
      domGetter: '#container',
      apps: [
        {
          name: 'react',
          entry: reactSubAppEntry,
        },
      ],
      beforeLoad: mockBeforeLoad,
    });

    const testName = 'react-unexisted';
    expect(GarfishInstance.appInfos[testName]).toBeUndefined;

    await GarfishInstance.loadApp(testName, {
      entry: reactSubAppEntry,
    });

    expect(mockBeforeLoad).toBeCalled();
    expect(mockBeforeLoad.mock.calls[0][0]).not.toBeUndefined;
    expect(mockBeforeLoad.mock.calls[0][0].name).toBe(testName);
    expect(mockBeforeLoad.mock.calls[0][0].entry).toBe(reactSubAppEntry);
  });

  it("loadApp before registered and `entry` don't be provided will toorw Error", async () => {
    GarfishInstance.run({
      domGetter: '#container',
      apps: [
        {
          name: 'react',
          entry: reactSubAppEntry,
        },
      ],
    });
    const testName = 'react-unexisted';
    expect(GarfishInstance.appInfos[testName]).toBeUndefined;

    GarfishInstance.loadApp(testName).catch((e) =>
      expect(e.message).toMatch(
        'Please provide the entry parameters or registered in advance of the app.',
      ),
    );
  });
});
