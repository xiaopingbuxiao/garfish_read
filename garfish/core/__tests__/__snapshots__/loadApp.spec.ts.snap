// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core: load process domGetter cssSelector 1`] = `
<div
  __garfishmockhtml__=""
>
  <div
    __garfishmockhead__=""
  >
    
    
    <title>
      vue sub app
    </title>
    
    
    <script
      type="garfish-config"
    >
      
      {
        "sandbox": {
          "noEntryScripts": ["scripts/no-entry.js"]
        }
      }
    
    </script>
    
    
    <!--&lt;script src="./scripts/no-entry.js"  execute by garfish&gt;&lt;/script&gt;-->
    
  
  </div>
  <div
    __garfishmockbody__=""
  >
    
    
    <div
      id="app"
    >
      <div
        id="hello-world"
      />
    </div>
    
    
    <!--&lt;script  execute by garfish&gt;
      if (typeof __GARFISH_EXPORTS__ !== 'undefined') {
        __GARFISH_EXPORTS__.provider = () =&gt; {
          return {
            render ({ dom , basename }) {
              let newContent = document.createElement('div');
              newContent.setAttribute('id', 'hello-world');
              dom.querySelector('#app').appendChild(newContent);
            },

            destroy ({ dom , basename }){
              dom.querySelector('#app').removeChild(dom.querySelector('#hello-world'));
            }
          }
        }
      }
    &lt;/script&gt;-->
    
  


  </div>
</div>
`;

exports[`Core: load process domGetter cssSelector delay 2000ms can work normally 1`] = `
<div
  __garfishmockhtml__=""
>
  <div
    __garfishmockhead__=""
  >
    
    
    <title>
      vue sub app
    </title>
    
    
    <script
      type="garfish-config"
    >
      
      {
        "sandbox": {
          "noEntryScripts": ["scripts/no-entry.js"]
        }
      }
    
    </script>
    
    
    <!--&lt;script src="./scripts/no-entry.js"  execute by garfish&gt;&lt;/script&gt;-->
    
  
  </div>
  <div
    __garfishmockbody__=""
  >
    
    
    <div
      id="app"
    >
      <div
        id="hello-world"
      />
    </div>
    
    
    <!--&lt;script  execute by garfish&gt;
      if (typeof __GARFISH_EXPORTS__ !== 'undefined') {
        __GARFISH_EXPORTS__.provider = () =&gt; {
          return {
            render ({ dom , basename }) {
              let newContent = document.createElement('div');
              newContent.setAttribute('id', 'hello-world');
              dom.querySelector('#app').appendChild(newContent);
            },

            destroy ({ dom , basename }){
              dom.querySelector('#app').removeChild(dom.querySelector('#hello-world'));
            }
          }
        }
      }
    &lt;/script&gt;-->
    
  


  </div>
</div>
`;

exports[`Core: load process domGetter is function 1`] = `
<div
  __garfishmockhtml__=""
>
  <div
    __garfishmockhead__=""
  >
    
    
    <title>
      vue sub app
    </title>
    
    
    <script
      type="garfish-config"
    >
      
      {
        "sandbox": {
          "noEntryScripts": ["scripts/no-entry.js"]
        }
      }
    
    </script>
    
    
    <!--&lt;script src="./scripts/no-entry.js"  execute by garfish&gt;&lt;/script&gt;-->
    
  
  </div>
  <div
    __garfishmockbody__=""
  >
    
    
    <div
      id="app"
    >
      <div
        id="hello-world"
      />
    </div>
    
    
    <!--&lt;script  execute by garfish&gt;
      if (typeof __GARFISH_EXPORTS__ !== 'undefined') {
        __GARFISH_EXPORTS__.provider = () =&gt; {
          return {
            render ({ dom , basename }) {
              let newContent = document.createElement('div');
              newContent.setAttribute('id', 'hello-world');
              dom.querySelector('#app').appendChild(newContent);
            },

            destroy ({ dom , basename }){
              dom.querySelector('#app').removeChild(dom.querySelector('#hello-world'));
            }
          }
        }
      }
    &lt;/script&gt;-->
    
  


  </div>
</div>
`;
